src/dmd/root/array.d

```diff
@@ -591,3 +591,81 @@ unittest
     static assert (!is(typeof(Array!(OtherStruct).init.sort())));
     static assert (!is(typeof(Array!(OtherStruct).init.sort!pred)));
 }
+
+/**
+ * Iterates the given array and calls the given callable for each element.
+ *
+ * Use this instead of `foreach` when the array may expand during iteration.
+ *
+ * Params:
+ *  callable = the callable to call for each element
+ *  array = the array to iterate
+ */
+template each(alias callable, T)
+if (is(typeof(callable(T.init))))
+{
+    void each(ref Array!T array)
+    {
+        // Do not use foreach, as the size of the array may expand during iteration
+        for (size_t i = 0; i < array.length; ++i)
+            callable(array[i]);
```

Comment:
Maybe this should allow early returns like `foreachDsymbol`?

Reply:
I didn't have a use case for that, but sure.

Reply:
Done.